src/hardware/voodoo.cpp

```diff
@@ -0,0 +1,7175 @@
+#define FBIINIT7_ENABLE_CHROMA_STUFF(val)	(((val) >> 19) & 1)		/* voodoo 2 only */
+#define FBIINIT7_CMDFIFO_PCI_TIMEOUT(val)	(((val) >> 20) & 0x7f)	/* voodoo 2 only */
+#define FBIINIT7_ENABLE_TEXTURE_BURST(val)	(((val) >> 27) & 1)		/* voodoo 2 only */
+
+#define TEXMODE_ENABLE_PERSPECTIVE(val)		(((val) >> 0) & 1)
+#define TEXMODE_MINIFICATION_FILTER(val)	(((val) >> 1) & 1)
+#define TEXMODE_MAGNIFICATION_FILTER(val)	(((val) >> 2) & 1)
+#define TEXMODE_CLAMP_NEG_W(val)			(((val) >> 3) & 1)
+#define TEXMODE_ENABLE_LOD_DITHER(val)		(((val) >> 4) & 1)
+#define TEXMODE_NCC_TABLE_SELECT(val)		(((val) >> 5) & 1)
+#define TEXMODE_CLAMP_S(val)				(((val) >> 6) & 1)
+#define TEXMODE_CLAMP_T(val)				(((val) >> 7) & 1)
+#define TEXMODE_FORMAT(val)					(((val) >> 8) & 0xf)
+#define TEXMODE_TC_ZERO_OTHER(val)			(((val) >> 12) & 1)
+#define TEXMODE_TC_SUB_CLOCAL(val)			(((val) >> 13) & 1)
+#define TEXMODE_TC_MSELECT(val)				(((val) >> 14) & 7)
+#define TEXMODE_TC_REVERSE_BLEND(val)		(((val) >> 17) & 1)
+#define TEXMODE_TC_ADD_ACLOCAL(val)			(((val) >> 18) & 3)
+#define TEXMODE_TC_INVERT_OUTPUT(val)		(((val) >> 20) & 1)
+#define TEXMODE_TCA_ZERO_OTHER(val)			(((val) >> 21) & 1)
+#define TEXMODE_TCA_SUB_CLOCAL(val)			(((val) >> 22) & 1)
+#define TEXMODE_TCA_MSELECT(val)			(((val) >> 23) & 7)
+#define TEXMODE_TCA_REVERSE_BLEND(val)		(((val) >> 26) & 1)
+#define TEXMODE_TCA_ADD_ACLOCAL(val)		(((val) >> 27) & 3)
+#define TEXMODE_TCA_INVERT_OUTPUT(val)		(((val) >> 29) & 1)
+#define TEXMODE_TRILINEAR(val)				(((val) >> 30) & 1)
+#define TEXMODE_SEQ_8_DOWNLD(val)			(((val) >> 31) & 1)
+
+#define TEXLOD_LODMIN(val)					(((val) >> 0) & 0x3f)
+#define TEXLOD_LODMAX(val)					(((val) >> 6) & 0x3f)
+#define TEXLOD_LODBIAS(val)					(((val) >> 12) & 0x3f)
+#define TEXLOD_LOD_ODD(val)					(((val) >> 18) & 1)
+#define TEXLOD_LOD_TSPLIT(val)				(((val) >> 19) & 1)
+#define TEXLOD_LOD_S_IS_WIDER(val)			(((val) >> 20) & 1)
+#define TEXLOD_LOD_ASPECT(val)				(((val) >> 21) & 3)
+#define TEXLOD_LOD_ZEROFRAC(val)			(((val) >> 23) & 1)
+#define TEXLOD_TMULTIBASEADDR(val)			(((val) >> 24) & 1)
+#define TEXLOD_TDATA_SWIZZLE(val)			(((val) >> 25) & 1)
+#define TEXLOD_TDATA_SWAP(val)				(((val) >> 26) & 1)
+#define TEXLOD_TDIRECT_WRITE(val)			(((val) >> 27) & 1)		/* Voodoo 2 only */
+
+#define TEXDETAIL_DETAIL_MAX(val)			(((val) >> 0) & 0xff)
+#define TEXDETAIL_DETAIL_BIAS(val)			(((val) >> 8) & 0x3f)
+#define TEXDETAIL_DETAIL_SCALE(val)			(((val) >> 14) & 7)
+#define TEXDETAIL_RGB_MIN_FILTER(val)		(((val) >> 17) & 1)		/* Voodoo 2 only */
+#define TEXDETAIL_RGB_MAG_FILTER(val)		(((val) >> 18) & 1)		/* Voodoo 2 only */
+#define TEXDETAIL_ALPHA_MIN_FILTER(val)		(((val) >> 19) & 1)		/* Voodoo 2 only */
+#define TEXDETAIL_ALPHA_MAG_FILTER(val)		(((val) >> 20) & 1)		/* Voodoo 2 only */
+#define TEXDETAIL_SEPARATE_RGBA_FILTER(val)	(((val) >> 21) & 1)		/* Voodoo 2 only */
+
+#define TREXINIT_SEND_TMU_CONFIG(val)		(((val) >> 18) & 1)
```

Comment:
Will this ever end???
It would be a massive work, but perhaps our bit_view would make this all go away...